File: backend/src/auto_stock_trading/api/market_data.py
from datetime import date
from typing import TYPE_CHECKING

from fastapi import APIRouter, HTTPException, status

from auto_stock_trading.api.market_data_models import (
    DailyBarResponse,
    DailyBarsResponse,
    InstrumentResponse,
    InstrumentsResponse,
    MinuteBarResponse,
    MinuteBarsResponse,
    QuoteResponse,
)

if TYPE_CHECKING:
    from auto_stock_trading.application.market_data import MarketDataReader
    from auto_stock_trading.domain.market_data.models import Instrument


def create_market_data_router(reader: MarketDataReader) -> APIRouter:
    router = APIRouter(prefix="/api/market-data/instruments", tags=["market-data"])

    async def instrument_list() -> InstrumentsResponse:
        results = await reader.instruments()
        return InstrumentsResponse(
            instruments=tuple(_instrument_response(result) for result in results)
        )

    async def instrument(symbol: str) -> InstrumentResponse:
        result = await reader.instrument(symbol)
        if result is None:
            raise HTTPException(status.HTTP_404_NOT_FOUND, "Instrument not found")
        return _instrument_response(result)

    async def quote(symbol: str) -> QuoteResponse:
        result = await reader.quote(symbol)
        if result is None:
            raise HTTPException(status.HTTP_404_NOT_FOUND, "Quote not found")
        return QuoteResponse(
            symbol=result.symbol,
            price=result.price,
            open_price=result.open_price,
            high_price=result.high_price,
            low_price=result.low_price,
            previous_close=result.previous_close,
            change=result.change,
            change_percent=result.change_percent,
            volume=result.volume,
            trading_value=result.trading_value,
            currency=result.currency,
            source=result.source,
            as_of=result.as_of,
            received_at=result.received_at,
        )

    async def daily_bars(
        symbol: str,
        start_date: date | None = None,
        end_date: date | None = None,
    ) -> DailyBarsResponse:
        if start_date is not None and end_date is not None and start_date > end_date:
            raise HTTPException(
                status.HTTP_422_UNPROCESSABLE_CONTENT,
                "start_date must not be after end_date",
            )
        results = await reader.daily_bars(symbol, start_date, end_date)
        await _ensure_known_instrument(reader, symbol, has_results=bool(results))
        bars = tuple(
            DailyBarResponse(
                trading_date=result.bar.trading_date,
                open_price=result.bar.open_price,
                high_price=result.bar.high_price,
                low_price=result.bar.low_price,
                close_price=result.bar.close_price,
                volume=result.bar.volume,
                trading_value=result.bar.trading_value,
                adjusted=result.bar.adjusted,
                correction_code=result.bar.correction_code,
                split_ratio=result.bar.split_ratio,
                source=result.bar.source,
                received_at=result.bar.received_at,
                finality=result.finality.value,
                confirmed_at=result.confirmed_at,
                version=result.version,
                valid_from=result.valid_from,
            )
            for result in results
        )
        return DailyBarsResponse(
            symbol=symbol,
            start_date=start_date,
            end_date=end_date,
            source=results[0].bar.source if results else None,
            bars=bars,
        )

    async def minute_bars(symbol: str, trading_date: date) -> MinuteBarsResponse:
        results = await reader.minute_bars(symbol, trading_date)
        await _ensure_known_instrument(reader, symbol, has_results=bool(results))
        bars = tuple(
            MinuteBarResponse(
                bar_started_at=result.bar.bar_started_at,
                open_price=result.bar.open_price,
                high_price=result.bar.high_price,
                low_price=result.bar.low_price,
                close_price=result.bar.close_price,
                volume=result.bar.volume,
                cumulative_trading_value=result.bar.cumulative_trading_value,
                source=result.bar.source,
                received_at=result.bar.received_at,
                finality=result.finality.value,
                confirmed_at=result.confirmed_at,
                version=result.version,
                valid_from=result.valid_from,
            )
            for result in results
        )
        return MinuteBarsResponse(
            symbol=symbol,
            trading_date=trading_date,
            source=results[0].bar.source if results else None,
            bars=bars,
        )

    router.add_api_route(
        "",
        instrument_list,
        methods=["GET"],
        description="수집 대상으로 등록된 종목 목록을 종목코드 순으로 반환한다.",
    )
    router.add_api_route("/{symbol}", instrument, methods=["GET"])
    router.add_api_route("/{symbol}/quote", quote, methods=["GET"])
    router.add_api_route("/{symbol}/daily-bars", daily_bars, methods=["GET"])
    router.add_api_route(
        "/{symbol}/minute-bars",
        minute_bars,
        methods=["GET"],
        description=(
            "검증된 시장 달력 세션 창 안의 비수정 1분봉 현재 버전을 반환한다. "
            "`pending`은 재관측 확정 전 값이며 전략 입력에 사용할 수 없다. "
            "`cumulative_trading_value`는 원본이 제공하는 당일 누적 거래대금이다."
        ),
    )
    return router


async def _ensure_known_instrument(
    reader: MarketDataReader,
    symbol: str,
    *,
    has_results: bool,
) -> None:
    if not has_results and await reader.instrument(symbol) is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Instrument not found")


def _instrument_response(result: Instrument) -> InstrumentResponse:
    return InstrumentResponse(
        country=result.country,
        exchange=result.exchange,
        symbol=result.symbol,
        product_type=result.product_type,
        currency=result.currency,
        name=result.name,
        english_name=result.english_name,
        listed_on=result.listed_on,
        delisted_on=result.delisted_on,
        trading_status=result.trading_status,
        source=result.source,
        source_as_of=result.source_as_of,
    )

File: frontend/src/lib/market-data.ts
import { z } from "zod";

const decimal = z.string().regex(/^-?\d+(\.\d+)?([Ee][+-]?\d+)?$/);
const nullableDecimal = decimal.nullable();
const isoDate = z.iso.date();
const isoDateTime = z.iso.datetime();

const instrumentSchema = z.strictObject({
  country: z.string().min(1),
  currency: z.string().min(1),
  delisted_on: isoDate.nullable(),
  english_name: z.string().nullable(),
  exchange: z.string().min(1),
  listed_on: isoDate.nullable(),
  name: z.string().min(1),
  product_type: z.enum(["stock", "etf"]),
  source: z.string().min(1),
  source_as_of: isoDate,
  symbol: z.string().min(1),
  trading_status: z.string().min(1),
});

const instrumentsSchema = z.strictObject({
  instruments: z.array(instrumentSchema).readonly(),
});

const quoteSchema = z.strictObject({
  as_of: isoDateTime,
  change: decimal,
  change_percent: decimal,
  currency: z.string().min(1),
  high_price: decimal,
  low_price: decimal,
  open_price: decimal,
  previous_close: decimal,
  price: decimal,
  received_at: isoDateTime,
  source: z.string().min(1),
  symbol: z.string().min(1),
  trading_value: decimal,
  volume: z.number().int().nonnegative(),
});

const dailyBarSchema = z.strictObject({
  adjusted: z.literal(false),
  close_price: decimal,
  confirmed_at: isoDateTime.nullable(),
  correction_code: z.string().nullable(),
  finality: z.enum(["pending", "confirmed"]),
  high_price: decimal,
  low_price: decimal,
  open_price: decimal,
  received_at: isoDateTime,
  source: z.string().min(1),
  split_ratio: nullableDecimal,
  trading_date: isoDate,
  trading_value: decimal,
  valid_from: isoDateTime,
  version: z.number().int().positive(),
  volume: z.number().int().nonnegative(),
});

const dailyBarsSchema = z.strictObject({
  bars: z.array(dailyBarSchema).readonly(),
  end_date: isoDate.nullable(),
  interval: z.literal("1d"),
  source: z.string().nullable(),
  start_date: isoDate.nullable(),
  symbol: z.string().min(1),
});

const corporateActionSchema = z.strictObject({
  action_key: z.uuid(),
  action_type: z.string().min(1),
  announced_at: isoDateTime.nullable(),
  announcement_date: isoDate,
  available_at: isoDateTime,
  cash_amount: nullableDecimal,
  corporate_action_id: z.uuid(),
  currency: z.string().nullable(),
  effective_date: isoDate.nullable(),
  ex_date: isoDate.nullable(),
  lifecycle: z.enum(["announced", "confirmed", "cancelled"]),
  payment_date: isoDate.nullable(),
  quality: z.enum(["pending", "verified", "conflict", "unsupported"]),
  received_at: isoDateTime,
  record_date: isoDate.nullable(),
  related_instrument_id: z.uuid().nullable(),
  share_multiplier: nullableDecimal,
  source: z.string().min(1),
  source_event_id: z.string().min(1),
  source_reference: z.string().min(1),
  subscription_price: nullableDecimal,
  superseded_at: isoDateTime.nullable(),
  time_precision: z.enum(["date", "minute", "second"]),
  valid_from: isoDateTime,
  version: z.number().int().positive(),
});

const corporateActionsSchema = z.strictObject({
  actions: z.array(corporateActionSchema).readonly(),
  end_date: isoDate.nullable(),
  include_history: z.boolean(),
  knowledge_cutoff_at: isoDateTime.nullable(),
  start_date: isoDate.nullable(),
  symbol: z.string().min(1),
});

const investorFlowSchema = z.strictObject({
  foreign_net_quantity: z.number().int(),
  foreign_net_value: z.number().int(),
  individual_net_quantity: z.number().int(),
  individual_net_value: z.number().int(),
  institution_net_quantity: z.number().int(),
  institution_net_value: z.number().int(),
  received_at: isoDateTime,
  trading_date: isoDate,
  version: z.number().int().positive(),
});

const investorFlowsSchema = z.strictObject({
  flows: z.array(investorFlowSchema).readonly(),
  quantity_unit: z.literal("share"),
  source: z.string().min(1),
  symbol: z.string().min(1),
  value_unit: z.literal("million_krw"),
});

const adjustedDatasetSchema = z.strictObject({
  action_version_hash: z.string().min(1),
  algorithm_version: z.string().min(1),
  dataset_id: z.uuid(),
  failure_code: z.string().nullable(),
  generated_at: isoDateTime,
  input_bar_version_hash: z.string().min(1),
  interval: z.literal("1d"),
  knowledge_cutoff_at: isoDateTime,
  method: z.enum(["split_adjusted", "total_return"]),
  price_cutoff_date: isoDate,
  range_start: isoDate,
  status: z.string().min(1),
  superseded_at: isoDateTime.nullable(),
  symbol: z.string().min(1),
});

const adjustedDailyBarSchema = z.strictObject({
  close_price: decimal,
  high_price: decimal,
  low_price: decimal,
  open_price: decimal,
  price_factor: decimal,
  source: z.string().min(1),
  source_bar_id: z.uuid(),
  source_bar_version: z.number().int().positive(),
  trading_date: isoDate,
  trading_value: decimal,
  volume: z.number().int().nonnegative(),
  volume_factor: decimal,
});

const appliedCorporateActionSchema = z.strictObject({
  action_key: z.uuid(),
  action_version: z.number().int().positive(),
  corporate_action_id: z.uuid(),
  event_date: isoDate,
  event_price_factor: decimal,
  event_volume_factor: decimal,
  source: z.string().min(1),
});

const adjustedDailyBarsSchema = z.strictObject({
  applied_actions: z.array(appliedCorporateActionSchema).readonly(),
  bars: z.array(adjustedDailyBarSchema).readonly(),
  dataset: adjustedDatasetSchema,
});

export type Instrument = z.infer<typeof instrumentSchema>;
export type Instruments = z.infer<typeof instrumentsSchema>;
export type Quote = z.infer<typeof quoteSchema>;
export type DailyBar = z.infer<typeof dailyBarSchema>;
export type DailyBars = z.infer<typeof dailyBarsSchema>;
export type CorporateAction = z.infer<typeof corporateActionSchema>;
export type CorporateActions = z.infer<typeof corporateActionsSchema>;
export type InvestorFlow = z.infer<typeof investorFlowSchema>;
export type InvestorFlows = z.infer<typeof investorFlowsSchema>;
export type AdjustedDailyBar = z.infer<typeof adjustedDailyBarSchema>;
export type AdjustedDailyBars = z.infer<typeof adjustedDailyBarsSchema>;

export const parseInstruments = (input: unknown): Instruments => instrumentsSchema.parse(input);
export const parseQuote = (input: unknown): Quote => quoteSchema.parse(input);
export const parseDailyBars = (input: unknown): DailyBars => dailyBarsSchema.parse(input);
export const parseCorporateActions = (input: unknown): CorporateActions =>
  corporateActionsSchema.parse(input);
export const parseInvestorFlows = (input: unknown): InvestorFlows =>
  investorFlowsSchema.parse(input);
export const parseAdjustedDailyBars = (input: unknown): AdjustedDailyBars =>
  adjustedDailyBarsSchema.parse(input);
